### PR TITLE
Added color support for the hud.

### PR DESCRIPTION
**Pull request by balzarot**
Now the hud honors the scr.color configuration and use colors instead of uppercase letters.
This should close Issue #3570

### CODE REVIEW

File: libr/cons/hud.c [hunk 1]
```diff
@@ -93,16 +93,36 @@ R_API char *r_cons_hud(RList *list, const char *prompt) {
...
+									p[j]=0;
+									r_cons_printf ("%s", p + last_color_change);
+									p[j]=tmp;
+								}
+								last_color_change=j+1;
+								r_cons_printf ("%s%c\x1b[0m", Color_GREEN,  p[j]);
+							}
+						}
+						r_cons_printf ("%s\n", p + last_color_change);
+					}
```
Comment (radare) [inline libr/cons/hud.c]: Wrog indent

**Comment by radare** [comment]
It would be better to reuse the color bit which is accessible from rcons instead of adding a var on every function and calling rconfig.get severa times

**Comment by radare** [comment]
And use-color should be const bool instead if int

File: libr/cons/hud.c (hunk 1 above)
Comment (balzarot) [inline libr/cons/hud.c]: Sorry, where is the wrong indentation ?

**Comment by balzarot** [comment]
I copied the way to access the color from the cmd_print file, but if there is a variable I can check in r_cons that would simplify it a lot (but then why so many commands do not use it ?).
Anyway, can you tell me which is the variable? I see many "int color" in various struct but I dont know what I should check..
Thanks a lot.

**Comment by radare** [comment]
it's in `(core->print->flags&R_PRINT_FLAGS_COLOR)` .... but i think we cant access this from RCons.. mmh

**Comment by radare** [comment]
Maybe RCons should have this information instead of rprint.. i would prefer not to have the same info in both modules..

File: libr/cons/hud.c (hunk 1 above)
Comment (ret2libc) [inline libr/cons/hud.c]: The else should be on the same as the brace. Please have a look here https://github.com/radare/radare2/blob/master/CONTRIBUTING.md#coding-style-guidelines for more info.

File: libr/cons/hud.c [hunk 2]
```diff
@@ -93,16 +93,36 @@ R_API char *r_cons_hud(RList *list, const char *prompt) {
...
 					p = strdup (pos);
-					for (j = 0; p[j]; j++) {
-						if (strchr (buf, p[j]))
-							p[j] = toupper ((unsigned char)p[j]);
+					if (usecolor) {
+						int last_color_change = 0;
+						char tmp;
+						r_cons_printf (" %c ", first? '-': ' ');
+						for (j = 0; p[j]; j++) {
+							if (strchr (buf, p[j])){
```
Comment (ret2libc) [inline libr/cons/hud.c]: In general, use spaces before braces (here and on the next line)

File: libr/cons/hud.c [hunk 3]
```diff
@@ -93,16 +93,36 @@ R_API char *r_cons_hud(RList *list, const char *prompt) {
...
+								if (j > last_color_change){
+									tmp = p[j];
+									p[j]=0;
```
Comment (ret2libc) [inline libr/cons/hud.c]: And spaces around `=`, `+`, etc.

```p[j] = 0;```

**Comment by ret2libc** [comment]
I think RCons should have the color bit, but that's probably for another issue/PR.
Also, the algorithm seems a little bit wrong to me, because it colours also chars that are not part of the pattern. IMO the whole pattern should be coloured inside the string (without considering the case of the letters) and not single chars like it is now. For example if you look for "str.", all 's' will be highlighted now, even if the the search looks for the whole pattern "str.". But this is also for another issue/PR, because it was this way also before. 

The patch looks good to me (apart from some style issues) ;)

**Comment by balzarot** [comment]
@ret2libc thanks for the comment and for the link to the style. I'll fix the code and squash the commits.
Regarding the algorithm, I noticed that it was strange but I copied what it was used before because maybe it was done on purpose like that. For instance, if you type: 

    > dyn|
    - 0x004002D8  sectioN..DYNsYm                                                      

also the first N and the last Y are matched. Which is weird.. 
If you want only exact substrings to be matched.. I can fix that both on the color and non-color version.
As I said.. I did not know if it was a bug or a feature :)

Finally, (@radare) should I switch from **r_config_get_i (core->config, "scr.color")**
to **core->print->flags&R_PRINT_FLAGS_COLOR** ?

**Comment by ret2libc** [comment]
(I think) it's a bug, but as said it was present even before :(
IMO exact substrings (ignoring the case) should be matched, because otherwise it looks a little bit weird. We can open a new issue for that.

**Comment by radare** [comment]
use color palete instead of hardcoded ansi codes

**Comment by radare** [comment]
or at least use the ansi constants Color_RED and such

**Comment by radare** [comment]
because of this `+r_cons_printf ("%s%c\x1b[0m", Color_GREEN,  p[j]);
`

**Comment by radare** [comment]
dat 1b[0m is Color_RESET and color-green doesnt needs to be concatenated with %s. it can be inlined in the const string to make printf's life easier

File: libr/cons/hud.c [hunk 4]
```diff
@@ -27,7 +27,7 @@ R_API char *r_cons_hud_file(const char *f) {
 	return NULL;
 }
 
-R_API char *r_cons_hud_string(const char *s) {
+R_API char *r_cons_hud_string(const char *s, const int usecolor) {
```
Comment (radare) [inline libr/cons/hud.c]: s/int/bool

**Comment by radare** [comment]
update copyright header from cons/hud.c

**Comment by balzarot** [comment]
Will do. 
I was teaching today.. but I will try to find some time tomorrow to fix everything.

**Comment by radare** [comment]
👍

**Comment by balzarot** [comment]
Ok, I have a working version. But before I update the PR I have a little question.
While I was working on the hud I decided to add some comments around (so others won't have to reverse the code again) and I noticed that at the beginning there is a todo list:

- [ ]  initial value
- [ ] default value (if null, use this)
- [ ] list of items filtered by user input
- [ ] execute command with info from row (display disasm or hexdump from off)
- [ ] we need a more complex data structure CSV?
   (string, key, command)
- [ ] commands can be passed to the hud callback
- [ ] commands: menu...

@radare can you tell me which are still valid and what do the other mean?

Finally, another little observation (which is not related to this PR but to the hud in general.
Now when the user leaves the hud it enters in visual mode.
However, it would not be more correct to go back to the previous mode? I often type V_ from
the commandline to quickly jump around.. and it would be great if then I get back to the prompt
again after I pick my target. What do you think?

**Comment by radare** [comment]
V is entering in visual mode, if you do V_ there's nothing to quit that visual mode. use V_q for this

**Comment by radare** [comment]
About the todo list i think we can completely remove it until we find the need to extend the current implementation, but i think it's good the way it is right now.

**Comment by radare** [comment]
Please update the PR

**Comment by balzarot** [comment]
updated

**Comment by radare** [comment]
Thanks . its merged now. i pushed this 360280a9b2a6b389e220fd8af4671a92e4b27597 to include some fixes

**Comment by radare** [comment]
thanks!

**Comment by balzarot** [comment]
Excellent. And thanks for the **V_q** trick. It does exactly what I needed (I still have a lot to learn about radare2)

**Pull request closed without merging**